vendor/k8s.io/kube-openapi/pkg/handler/handler.go

```diff
@@ -55,13 +56,40 @@ type OpenAPIService struct {
 
 	lastModified time.Time
 
-	specBytes []byte
-	specPb    []byte
-	specPbGz  []byte
+	jsonCache  cache
+	protoCache cache
+}
+
+type cache struct {
+	BuildCache func() ([]byte, error)
+	once       sync.Once
+	bytes      []byte
+	etag       string
+	err        error
+}
 
-	specBytesETag string
-	specPbETag    string
-	specPbGzETag  string
+func (c *cache) Get() ([]byte, string, error) {
+	c.once.Do(func() {
+		bytes, err := c.BuildCache()
+		// if there is an error updating the cache, there can be situations where
+		// c.bytes contains a valid value (carried over from the previous update)
+		// but c.err is also not nil; the cache user is expected to check for this
+		c.err = err
+		if c.err == nil {
+			// don't override previous spec if we had an error
+			c.bytes = bytes
+			c.etag = computeETag(c.bytes)
```

Comment:
It's pretty twisty that Get() mutates *cache, and that calls to Get() from getSwaggerBytes() and getSwaggerPbBytes() are only protected by a read-lock.

It's not clear from the outside that cache#Get() and cache#New() are unsafe to call concurrently... and the calls to Get and New only happens to be safe because they're guarded by locks (which were primarily intended to protect reads/writes to the fields in OpenAPIService).

Not blocking for this PR, but probably worth a follow-up in kube-openapi master to clarify/clean that up

Reply:
Good catch! Let's fix that up.